### ISSUE TITLE
Issue mocking Shell class in Cake 2.x

### ISSUE THREAD

**Issue by garethellis36** [cakephp/cakephp#8678]
I have some unit tests for custom shell classes. In order to avoid actually outputting anything to the console while the test is running, I have to provide mocks of `ConsoleOutput` and `ConsoleInput` to the shell constructor.

However, the constructor of the parent `Shell` class invokes `Shell::_useLogger()` which in turn sets console logging to happen using the mock stdout and stderr instances.

This means that if you run these tests as part of a larger test suite, you get no error output in the console for any errors which occur in tests run after the shell test.

I have come up with two possible workarounds but neither feels particularly great:

1. Use PHPUnit to create a partial mock and stub the `_useLogger()` method or disable the original constructor. I don't like this because I don't like having to use a partial mock just for this.
1. Add a function to bootstrap called `setConsoleLogger()` which resets the console logger to the correct one - this has to be called after every instance of invoking a child instance of `Shell` using a mock `ConsoleOutput`. I don't like this because it's boilerplate and harder to maintain.

Thoughts?

**Comment by markstory** [comment]
Why is passing mocks into the loggers problematic? That's the approach we use for testing the core shells and it seems to work well.

**Comment by garethellis36** [comment]
Is that because `ShellTest` uses a partial mock which stubs the `_useLogger()` method? Without a partial mock, the `_useLogger()` method replaces the stream used by the `stderr` and `stdout` log configurations to use the mock `ConsoleOutput`, so nothing gets output by errors in subsequent tests.

```php
//Shell.php
	protected function _useLogger($enable = true) {
		if (!$enable) {
			CakeLog::drop('stdout');
			CakeLog::drop('stderr');
			return;
		}
		CakeLog::config('stdout', array(
			'engine' => 'Console',
			'types' => array('notice', 'info'),
			'stream' => $this->stdout,
		));
		CakeLog::config('stderr', array(
			'engine' => 'Console',
			'types' => array('emergency', 'alert', 'critical', 'error', 'warning', 'debug'),
			'stream' => $this->stderr,
		));
	}
```

**Comment by markstory** [comment]
For the core tests, and for the projects I work on. Not having log output in the tests is a positive, not a negative :smile:

**Comment by garethellis36** [comment]
What about something like this? Simply makes sure that an existing configuration doesn't get overwritten:

```php
//Shell.php
	protected function _useLogger($enable = true) {
		if (!$enable) {
			CakeLog::drop('stdout');
			CakeLog::drop('stderr');
			return;
		}
               $configured = CakeLog::configured();
               if (!in_array("stdout", $configured)) {
		CakeLog::config('stdout', array(
			'engine' => 'Console',
			'types' => array('notice', 'info'),
			'stream' => $this->stdout,
		));
              }
              if (!in_array("stderr", $configured)) {
		CakeLog::config('stderr', array(
			'engine' => 'Console',
			'types' => array('emergency', 'alert', 'critical', 'error', 'warning', 'debug'),
			'stream' => $this->stderr,
		));
             }
	}
```

**Comment by markstory** [comment]
Not overwriting loggers is a reasonable option.

**Comment by markstory** [comment]
Could you open a pull request with the proposed changes?

**Comment by garethellis36** [comment]
I've started this. https://github.com/cakephp/cakephp/compare/2.next...garethellis36:dont-override-shell-loggers

A couple of issues:

1. One of the tests in `ShellTest` outputs `Error: all your base are belong to us` to `stderr`. I can't find where this is coming from. I guess I would need to call `CakeLog::disable('stderr')` in that test to stop it outputting. 
2. I created my branch from the `2.8` branch but i don't seem able to submit a PR against that branch. When I select 2.8, GitHub tells me "nothing to compare". I can only submit a PR against `2.next` - is that right?
3. One of the tests in `ShellTest` was failing before I changed anything. I'm assuming this is a configuration issue and that it will pass OK on CI...
